### PR TITLE
Improve the function pointer docs

### PR DESCRIPTION
**Pull request by notriddle**
This is #97842 but for function pointers instead of tuples. The concept is basically the same.

* Reduce duplicate impls; show `fn (T₁, T₂, …, Tₙ)` and include a sentence saying that there exists up to twelve of them.
* Show `Copy` and `Clone`.
* Show auto traits like `Send` and `Sync`, and blanket impls like `Any`.

https://notriddle.com/notriddle-rustdoc-test/std/primitive.fn.html

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Just looked at your hosted docs and saw this:

Screenshot from 2022-07-18 11-14-36

Not sure if it comes from this PR or not though. If not, I'll resume the review and open an issue. :)

File: library/core/src/primitive_docs.rs [hunk 1]
```diff
@@ -1441,11 +1441,16 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
-/// ### Traits
+/// ### Trait implementations
 ///
-/// Function pointers implement the following traits:
+/// In this documentation the shorthand `fn (T₁, T₂, …, Tₙ)` is used to represent non-variadic
+/// function pointers of varying length. Note that this is a convenience notation to avoid
+/// repepitive documentation, not valid Rust syntax.
```
Comment (GuillaumeGomez) [inline library/core/src/primitive_docs.rs:1448]:
```suggestion
/// repetitive documentation, not valid Rust syntax.
```

File: library/core/src/ptr/mod.rs [hunk 2]
```diff
@@ -1896,7 +1938,7 @@ macro_rules! fnptr_impls_args {
 }
 
 fnptr_impls_args! {}
-fnptr_impls_args! { A }
+fnptr_impls_args! { T }
```
Comment (GuillaumeGomez) [inline library/core/src/ptr/mod.rs:1941]: Why this change?

File: library/std/src/primitive_docs.rs [hunk 3]
```diff
@@ -1441,11 +1441,16 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
-/// ### Traits
+/// ### Trait implementations
 ///
-/// Function pointers implement the following traits:
+/// In this documentation the shorthand `fn (T₁, T₂, …, Tₙ)` is used to represent non-variadic
+/// function pointers of varying length. Note that this is a convenience notation to avoid
+/// repepitive documentation, not valid Rust syntax.
```
Comment (GuillaumeGomez) [inline library/std/src/primitive_docs.rs:1448]:
```suggestion
/// repetitive documentation, not valid Rust syntax.
```

File: src/librustdoc/html/format.rs [hunk 4]
```diff
@@ -1165,10 +1165,43 @@ impl clean::Impl {
...
             if let clean::Type::Tuple(types) = &self.for_ &&
                 let [clean::Type::Generic(name)] = &types[..] &&
-                (self.kind.is_tuple_variadic() || self.kind.is_auto()) {
+                (self.kind.is_fake_variadic() || self.kind.is_auto()) {
                 // Hardcoded anchor library/core/src/primitive_docs.rs
                 // Link should match `# Trait implementations`
                 primitive_link_fragment(f, PrimitiveType::Tuple, &format!("({name}₁, {name}₂, …, {name}ₙ)"), "#trait-implementations-1", cx)?;
+            } else if let clean::BareFunction(bare_fn) = &self.for_ &&
+                let [clean::Argument { type_: clean::Type::Generic(name), .. }] = &bare_fn.decl.inputs.values[..] &&
+                (self.kind.is_fake_variadic() || self.kind.is_auto()) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/format.rs:1174]: I'm surprised rustfmt didn't put the `{` on a new line. Took me a while to understand what I was reading. ^^'

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks! I'd just like to have someone from the compiler to confirm the changes on their side before approving it though. I'll let you ping someone from their team.

File: src/librustdoc/html/format.rs [hunk 5]
```diff
@@ -1165,10 +1165,45 @@ impl clean::Impl {
...
+            {
+                // Hardcoded anchor library/core/src/primitive_docs.rs
+                // Link should match `# Trait implementations`
+                if f.alternate() {
+                    write!(
+                        f,
+                        "{hrtb:#}{unsafety}{abi:#}",
+                        hrtb = bare_fn.print_hrtb_with_space(cx),
+                        unsafety = bare_fn.unsafety.print_with_space(),
+                        abi = print_abi_with_space(bare_fn.abi),
```
Comment (petrochenkov) [inline src/librustdoc/html/format.rs:1185]: The `hrtb/unsafety/abi = ...` part can be moved out of the `if f.alternate()` condition.

**Comment by petrochenkov** [comment]
r=me after squashing commits.
@rustbot author

**Comment by pnkfelix** [comment]
The regressions above are all in doc generation, and they are all minor.

@rustbot label: +perf-regression-triaged
